Allow filtering by organisation for harvest_source_list

1pretz1 opened the pull request:
This commit allows users of the harvest_source_list endpoint to
return harvest sources by a specific organisation by using a query param
`organization_id`. Also this endpoint won't
return more than 100 harvest sources as we were having timing out issues
due to the amount being previously returned (500+).

I've also removed the check_access method as this currently wasn't doing
anything as you didn't need to be authorized to access this endpoint.

Trello:
https://trello.com/c/iAWEYNOv/1946-5-dgu-add-ability-for-harvestsourcelist-to-filter-by-organisation%F0%9F%8D%90

Zendesk:
https://govuk.zendesk.com/agent/tickets/4034244

1pretz1 committed:
Allow filtering by organisation for harvest_source_list

This commit allows users of the harvest_source_list endpoint to
return harvest sources by a specific organisation by using a query param
`organization_id`. Also this endpoint won't
return more than 100 harvest sources as we were having timing out issues
due to the amount being previously returned (500+).

I've also removed the check_access method as this currently wasn't doing
anything as you didn't need to be authorized to access this endpoint.

Trello:
https://trello.com/c/iAWEYNOv/1946-5-dgu-add-ability-for-harvestsourcelist-to-filter-by-organisation%F0%9F%8D%90

Zendesk:
https://govuk.zendesk.com/agent/tickets/4034244

kentsanggds commented on ckanext/harvest/logic/action/get.py:
```diff
@@ -361,17 +361,26 @@ def harvest_log_list(context, data_dict):
 
+    organization_id = request.params.get('organization_id')
+    if harvest_source_list:
+        limit = int(os.environ.get("HARVEST_SOURCE_LIST_LIMIT", "100"))
```
Rather than using env vars, I think the convention in ckan is to get the setting from config object - https://github.com/alphagov/ckanext-harvest/blob/master/ckanext/harvest/queue.py#L39

1pretz1 replied:
Okay sure, I've used config objects

kentsanggds commented on ckanext/harvest/tests/nose/test_action.py:
```diff
@@ -258,6 +259,73 @@ def test_create(self):
+        assert 1 is len(results)
+        assert_equal(organization["id"], result_organization_id)
+
+    def test_list_without_organization_and_limit(self):
```
Can we also have a test where there is no limit set, as this could be the case if other devs don't set the HARVEST_SOURCE_LIST_LIMIT config

1pretz1 replied:
Sure, I've added a test and re-organised them slightly so all test cases should be covered

kentsanggds commented on ckanext/harvest/logic/action/get.py:
```diff
@@ -1,11 +1,13 @@
+import os
 from sqlalchemy import or_
-from ckan.model import User
+from ckan.model import (User, Package)
```
You don't need brackets around the import here

kentsanggds commented on ckanext/harvest/logic/action/get.py:
```diff
@@ -124,9 +126,7 @@ def harvest_source_list(context, data_dict):
-    check_access('harvest_source_list', context, data_dict)
-
-    sources = _get_sources_for_user(context, data_dict)
+    sources = _get_sources_for_user(context, data_dict, harvest_source_list=True)
```
Rather than pass in a flag for where the call is coming from, perhaps pass in the organisation_id, as it appears that the other call which uses `_get_sources_for_user` is called from the command line so potentially might not have a `request` object. Limit could also be passed in as another parameter, defaulting to a 100 if it is not set.

1pretz1 replied:
Good spot! I've updated the PR as per your suggestions which I actually think makes the code look neater too.

kentsanggds reviewed (changes requested):
A few changes to improve the code, and move some logic around.

1pretz1 commented:
Thanks for the review @kentsanggds, I've updated the PR accordingly.

kentsanggds commented on ckanext/harvest/tests/nose/test_action.py:
```diff
@@ -258,6 +259,95 @@ def test_create(self):
+        assert 2 is len(results)
+
+    def test_list_with_limit(self):
+        config["ckan.harvest.harvest_source_limit"] = 1
```
Rather than set the config to 1, you should be able to patch the value like this - 

https://github.com/alphagov/ckanext-harvest/blob/master/ckanext/harvest/tests/nose/harvesters/test_base.py#L25-L26

The actual patch code is something like this - 

```
    @patch.dict('ckanext.harvest.logic.action.get.config',
                {'ckan.harvest.harvest_source_limit': 1})
```

1pretz1 replied:
Sure, didn't know you could mock in this way.

kentsanggds commented on ckanext/harvest/tests/nose/test_action.py:
```diff
@@ -258,6 +259,95 @@ def test_create(self):
+        assert response.json['success'] is True
+        assert 1 is len(results)
+
+        config["ckan.harvest.harvest_source_limit"] = None
```
This line won't be needed if patching the config.

kentsanggds reviewed (changes requested):
A minor change to `patch` rather than set the config setting in the test code.

kentsanggds commented:
One more thing, it's probably worth mentioning the `harvest_source_limit` in the README so that other devs will know about it without having to look through the code.

1pretz1 committed:
Add documentation around harvest_source_list API

This adds some documentation around the harvest_source_list API
endpoint. This is meant to be the start of an endpoint list for ckanext-harvest
which eventually should be extended to include all the other endpoints.

1pretz1 commented:
I've updated the PR which should be ready for another look @kentsanggds!

kentsanggds commented on README.rst:
```diff
@@ -885,6 +885,21 @@ Here are some common errors and solutions:
+
+This endpoint will return all the harvest sources in CKAN with a default limit
+of 100 items. The limit can be set to a bespoke value in the config for ckan
+under `harvest_source_limit`.
```
MIght be better to actually write out what the config setting is, will make it clearer for devs.

kentsanggds reviewed (changes requested):
Minor change to show the config setting rather than mention it in passing.
